### PR TITLE
str: common logic for replace & replacen

### PR DESCRIPTION
**Pull request by ljedrz**
The logic of `str::replace` and `str::replacen` is almost identical - we can use a common helper function. The initial capacity of the resulting string was different (`0` vs. `32`), but I propose a different approach: use the same capacity as in the original string; this will neither reallocate too often (or not at all if the substitute's length is the same as the substituted bit's), nor will it overallocate for short strings. It will also be an exact fit in case of no pattern hits.

### CODE REVIEW

**Comment by kennytm** [comment]
I'm going to rewrite these functions for #56345, would you mind if I just incorporate these changes directly into the new PR?

**Comment by ljedrz** [comment]
Sure, I don't mind at all  :ok_hand:.

**Comment by kennytm** [comment]
OK I'll close this when that PR is submitted.

File: src/liballoc/str.rs [hunk 1]
```diff
@@ -622,3 +605,16 @@ impl str {
 pub unsafe fn from_boxed_utf8_unchecked(v: Box<[u8]>) -> Box<str> {
     Box::from_raw(Box::into_raw(v) as *mut str)
 }
+
+fn replace_helper<'a, P: Pattern<'a>>(s: &'a str, pat: P, to: &str, lim: Option<usize>) -> String {
```
Comment (Centril) [inline src/liballoc/str.rs]: A comment would be nice... :)

File: src/liballoc/str.rs [hunk 2]
```diff
@@ -622,3 +605,16 @@ impl str {
...
+    let mut result = String::with_capacity(s.len());
+    let mut last_end = 0;
+    let limit = if let Some(limit) = lim { limit } else { s.len() };
+    for (start, part) in s.match_indices(pat).take(limit) {
+        result.push_str(unsafe { s.get_unchecked(last_end..start) });
```
Comment (Centril) [inline src/liballoc/str.rs]: And a comment explaining why this is safe as well...

File: src/liballoc/str.rs [hunk 3]
```diff
@@ -622,3 +605,16 @@ impl str {
...
+        result.push_str(to);
+        last_end = start + part.len();
+    }
+    result.push_str(unsafe { s.get_unchecked(last_end..s.len()) });
```
Comment (Centril) [inline src/liballoc/str.rs]: and here...

**Review by Centril — CHANGES_REQUESTED** [review]

**Comment by Dylan-DPC** [comment]
@ljedrz can you update this PR with the changes requested above?

**Comment by ljedrz** [comment]
@Dylan-DPC actually @kennytm is interested in intercepting this change (see https://github.com/rust-lang/rust/pull/56464#issuecomment-443758799); that's why I'm not updating this PR - it will be closed when it becomes a part of https://github.com/rust-lang/rust/issues/56345.

**Comment by Dylan-DPC** [comment]
Thanks. Marking it as blocked.

**Comment by Centril** [comment]
Closing meanwhile.

**Pull request closed without merging**